sw/device/silicon_creator/lib/drivers/watchdog.c

```diff
@@ -5,42 +5,79 @@
 #include "sw/device/silicon_creator/lib/drivers/watchdog.h"
 
 #include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/silicon_creator/lib/base/sec_mmio.h"
+#include "sw/device/silicon_creator/lib/drivers/lifecycle.h"
+#include "sw/device/silicon_creator/lib/drivers/otp.h"
 
 #include "aon_timer_regs.h"
 #include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "otp_ctrl_regs.h"
 #include "pwrmgr_regs.h"
 
 enum {
   kBase = TOP_EARLGREY_AON_TIMER_AON_BASE_ADDR,
   kPwrMgrBase = TOP_EARLGREY_PWRMGR_AON_BASE_ADDR,
-  // The AON domain is always clocked at 200 Hz.
-  // TODO(lowRISC/opentitan#7385): update this after the AON frequency is
-  // formally defined in the project.
-  kAonTimerRate = 200000,
+
+  kCtrlEnable = 1 << AON_TIMER_WDOG_CTRL_ENABLE_BIT,
+  kCtrlDisable = 0 << AON_TIMER_WDOG_CTRL_ENABLE_BIT,
 };
 
-void watchdog_init(uint32_t timeout_ms) {
+void watchdog_init(lifecycle_state_t lc_state) {
+  // Disable the watchdog bite when in test and RMA lifecycle states.
+  hardened_bool_t enable = kHardenedBoolTrue;
+  switch (lc_state) {
+    case kLcStateTest:
+      enable = kHardenedBoolFalse;
+      HARDENED_CHECK_EQ(lc_state, kLcStateTest);
+      break;
+    case kLcStateRma:
+      enable = kHardenedBoolFalse;
+      HARDENED_CHECK_EQ(lc_state, kLcStateRma);
+      break;
+    default:
+      break;
+  }
+
+  uint32_t threshold =
+      otp_read32(OTP_CTRL_PARAM_ROM_WATCHDOG_BITE_THRESHOLD_CYCLES_OFFSET);
+  watchdog_configure(threshold, enable);
+}
+
+void watchdog_configure(uint32_t threshold, hardened_bool_t enable) {
   // Tell pwrmgr we want watchdog reset events to reset the chip.
-  abs_mmio_write32(
+  sec_mmio_write32(
       kPwrMgrBase + PWRMGR_RESET_EN_REG_OFFSET,
       bitfield_bit32_write(
           0, kTopEarlgreyPowerManagerResetRequestsAonTimerAonAonTimerRstReq,
           true));
   abs_mmio_write32(kPwrMgrBase + PWRMGR_CFG_CDC_SYNC_REG_OFFSET, 1);
 
-  // Disable the watchdog before reconfiguring it.
-  abs_mmio_write32(kBase + AON_TIMER_WDOG_CTRL_REG_OFFSET, 0);
-  abs_mmio_write32(kBase + AON_TIMER_WKUP_CTRL_REG_OFFSET, 0);
-  // Configure the watchdog to bite at the requested timeout.
-  abs_mmio_write32(kBase + AON_TIMER_WKUP_COUNT_REG_OFFSET, 0);
+  // Set the watchdog bite and bark thresholds.
   abs_mmio_write32(kBase + AON_TIMER_WDOG_COUNT_REG_OFFSET, 0);
-  abs_mmio_write32(kBase + AON_TIMER_WKUP_THOLD_REG_OFFSET, UINT32_MAX);
   abs_mmio_write32(kBase + AON_TIMER_WDOG_BARK_THOLD_REG_OFFSET, UINT32_MAX);
-  abs_mmio_write32(kBase + AON_TIMER_WDOG_BITE_THOLD_REG_OFFSET,
-                   timeout_ms * (kAonTimerRate / 1000));
-  if (timeout_ms) {
-    abs_mmio_write32(kBase + AON_TIMER_WDOG_CTRL_REG_OFFSET, 1);
+  sec_mmio_write32(kBase + AON_TIMER_WDOG_BITE_THOLD_REG_OFFSET, threshold);
```

Comment:
we should disallow configuration if the threshold is set to 0. Maybe just compare against a reasonable threshold.

Reply:
@moidx what sort of action should we take in that scenario? round up to a minimum value?

Reply:
#10809